### ISSUE TITLE
Issue: Difference in no-circular analysis when run on Windows vs Linux

### ISSUE THREAD

**Issue by binary64** [sverweij/dependency-cruiser#498]
On my Windows machine, I do this just fine:
```
depcruise libs apps --config .dependency-cruiser.js --output-type baseline --output-to .dependency-cruiser-known-violations.json
depcruise libs apps --config .dependency-cruiser.js --ignore-known # <-- exit code 0
```

But the exact same code on Linux:
```
docker run --rm -it -v "c:/Users/User/.ssh/id_rsa:/root/id_rsa:ro" node:16-buster bash -c "mkdir -p /root/.ssh && cp /root/id_rsa /root/.ssh/ && chmod 0600 /root/.ssh/id_rsa && git clone ... && cd monorepo && yarn && yarn depcruise libs apps --config .dependency-cruiser.js --ignore-known"
```

Outputs this:
```
$ depcruise libs apps --config .dependency-cruiser.js --ignore-known

  error no-circular: apps/web/components/Users/ListView/ListView.tsx →
      apps/web/pages/users/index.tsx →
      apps/web/components/Users/index.ts →
      apps/web/components/Users/ListView/ListView.tsx

✖ 1 dependency violations (1 errors, 0 warnings). 6375 modules, 13919 dependencies cruised.

error Command failed with exit code 1.
```

### Expected Behavior

The violations should perfectly match whether you depcruise on Windows or Linux.

### Context

There are like 850 violations in my `.dependency-cruiser-known-violations.json` file and the discrepency here is just ONE difference. Typical! 🗡️ 

This is blocking me rolling it out to CI

### Your Environment

- Version used: 10.3.0-beta-4
- Node version: 16.5 for windows, 16-buster docker tag for linux

**Comment by sverweij** [comment]
@binary64 need a little more information here. I assume you've used the known violations you generated on windows as input for the command on linux?

I can _guess_ what is going on (see below) - but will need a minimal reproduction example to be sure.

`<conjecture>`
You might've noticed that when there's a cycle e.g. src/cycle-1.js → src/cycle-2.js → src/cycle-3.js → src/cycle-1.js dependency-cruiser in its error summary on reports _one_ violation (even there's 3 in its data model). This is because it's the same cycle and reporting these things multiple times would make the output quite noisy, even for a simple cycle of 3:

src/cycle-1.js → src/cycle-2.js → src/cycle-3.js → src/cycle-1.js 
src/cycle-2.js → src/cycle-3.js → src/cycle-1.js → src/cycle-2.js 
src/cycle-3.js → src/cycle-1.js → src/cycle-2.js → src/cycle-3.js

So dependency-cruiser picks one of these (the first it encounters). However, the order currently depends on what the file system fed it (which in turn depends on order they were written/ last accessed/ the locale of the OS/ the moon's current tide). Until now this wasn't really a problem, but with the ignore-known feature (which is still in beta) it might be. If the known-violations has the first and the encountered violation has the third it'll bump into the above situation.

If this is indeed the root-cause probably be sorted by ordering the violations before uniq'ing them.
`</conjecture>`

To prove the conjecture you could generate the known-violations file twice - once on windows and once on linux. If the circular dependency differs between both we've found the culprit (and otherwise we'd have to search a little longer).

**Comment by sverweij** [comment]
@binary64 could you try with `dependency-cruiser@10.3.0-beta-7`? That one sorts violations before it uniq's them instead of after, which might solve the issue when the above conjecture holds true. 

=> When you test this could you first re-generate the known-errors on windows again before running (with that same known-error file) on linux?

**Comment by binary64** [comment]
Just tried beta-7 but it's a lot worse - ~250 new violations on Linux, compared with Windows. I'll investigate a bit more next week

**Comment by sverweij** [comment]
@binary64 - hmm that's odd. I don't think they're new violations b.t.w. as explained earlier - it's probably the same violations but noted down a little differently. 

=> did you re-generated the known errors on windows (and committed them) with beta 7 and then ran with _that_ file on linux (also with beta 7)?

If that didn't happen than the larger difference between windows and linux is easily explained. 

Looking forward to a minimal reproduction sample!

**Comment by sverweij** [comment]
hi @binary64 I've tested the conjecture from a few comments ago and it does seem to hold up. Take three modules that are related like so:

<img width="173" alt="image" src="https://user-images.githubusercontent.com/4822597/133666892-123aae7e-0235-4d32-87bd-8329cdb24bc6.png">

conjecture: when we feed dependency-cruiser first with src/cycle-1.js or first with src/cycle-2.js these two runs will ...
- yield the same violations
- represent the violations differently; the cycles will  start on cycle-1.js and cycle-2.js respectively


### result 
(both with with dependency-cruiser@10.2.0 and dependency-cruiser@10.3.0-beta-6). The predicted difference is indeed there:

```sh
$ depcruise src/cycle-1.js --config --output-type err

  error no-circular: src/cycle-1.js → 
      src/cycle-2.js →
      src/cycle-3.js →
      src/cycle-1.js

✖ 1 dependency violations (1 errors, 0 warnings). 3 modules, 3 dependencies cruised.

$ depcruise src/cycle-2.js --config --output-type err

  error no-circular: src/cycle-2.js → 
      src/cycle-3.js →
      src/cycle-1.js →
      src/cycle-2.js

✖ 1 dependency violations (1 errors, 0 warnings). 3 modules, 3 dependencies cruised.
```

### after applying reproducible sorting:

The next version of dependency-cruiser goes through some length to make ordering consistent, regardless the quirks of the filesystem, locales and whatnot (part of PR #494). With that version the same violation is represented consistently. This will also make that --ignore-known will be able to match them.

```sh
$ depcruise src/cycle-1.js --config --output-type err

  error no-circular: src/cycle-1.js → 
      src/cycle-2.js →
      src/cycle-3.js →
      src/cycle-1.js

✖ 1 dependency violations (1 errors, 0 warnings). 3 modules, 3 dependencies cruised.

$ depcruise src/cycle-2.js --config --output-type err

  error no-circular: src/cycle-1.js → 
      src/cycle-2.js →
      src/cycle-3.js →
      src/cycle-1.js

✖ 1 dependency violations (1 errors, 0 warnings). 3 modules, 3 dependencies cruised.
```

**Comment by sverweij** [comment]
Hi @binary64 I've reproduced the problem you encountered with beta-7 on a private repo - no need to even switch platforms for it. Root cause is clear, as is a likely fix. I might have some time to implement it in the upcoming days.

**Comment by sverweij** [comment]
@binary64 the fix is up on part of  dependency-cruiser@10.3.1 - which has just been published on npm.

There were two root-causes. The first one was in known-violations functionality that needed special treatment for cyclic dependencies. The other (and more importantly) one was in the summarising functionality. That was too enthusiastic de-duplicating circular dependencies. Because the baseline reporter uses the summary the .dependency-cruiser-known-violations.json would contain only a subset of cyclic violations.

This means that in order to use it you'll have to _depcruise-baseline_ first before employing next steps.
